apps/src/netsim/NetSimTable.js

```diff
@@ -0,0 +1,186 @@
+/**
+ * Copyright 2015 Code.org
+ * http://code.org/
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/* jshint
+ funcscope: true,
+ newcap: true,
+ nonew: true,
+ shadow: false,
+ unused: true,
+
+ maxlen: 90,
+ maxparams: 3,
+ maxstatements: 200
+ */
+'use strict';
+
+var _ = require('../utils').getLodash();
+var ObservableEvent = require('./ObservableEvent');
+
+/**
+ * Maximum time (in milliseconds) that tables should wait between full cache
+ * updates from the server.
+ * @type {number}
+ */
+var POLLING_DELAY_MS = 5000;
+
+/**
+ * Wraps the app storage table API in an object with local
+ * cacheing and callbacks, which provides a notification API to the rest
+ * of the NetSim code.
+ * @param {!SharedTable} storageTable - The remote storage table to wrap.
+ * @constructor
+ */
+var NetSimTable = module.exports = function (storageTable) {
+  /**
+   * Actual API to the remote shared table.
+   * @type {SharedTable}
+   * @private
+   */
+  this.remoteTable_ = storageTable;
+
+
+  /**
+   * Event that fires when full table updates indicate a change,
+   * when rows are added, or when rows are removed, or when rows change.
+   * @type {ObservableEvent}
+   */
+  this.tableChangeEvent = new ObservableEvent();
+
+  /**
+   * Store table contents locally, so we can detect when changes occur.
+   * @type {Object}
+   * @private
+   */
+  this.cache_ = {};
+
+  /**
+   * Unix timestamp for last time this table's cache contents were fully
+   * updated.  Used to determine when to poll the server for changes.
+   * @type {number}
+   * @private
+   */
+  this.lastFullUpdateTime_ = 0;
+};
+
+NetSimTable.prototype.readAll = function (callback) {
+  var self = this;
```

Comment:
This seems like a good place to use `.bind(this)` but no need to change if you prefer `self`.